cmd/zoekt-sourcegraph-indexserver/queue.go

```diff
@@ -124,7 +129,112 @@ func (q *Queue) Iterate(f func(*IndexOptions)) {
 	q.mu.Lock()
 	defer q.mu.Unlock()
 	for _, item := range q.items {
-		f(&item.opts)
+		f(&item.Opts)
+	}
+}
+
+// debugEncodeSortedGobStream gob-encodes all items contained in the queue
+// (sorted by priority) to the provided writer.
+func (q *Queue) debugEncodeSortedGobStream(w io.Writer) error {
+	enc := gob.NewEncoder(w)
+
+	var err error
+	q.debugIteratedOrdered(func(item *queueItem) {
+		if err != nil {
+			return
+		}
+
+		err = enc.Encode(streamQueueReply{
+			Type: streamReplyQueueItem,
+			Item: item,
+		})
+	})
+
+	if err != nil {
+		reply := streamQueueReply{
+			Type:        streamReplyError,
+			ErrorString: fmt.Sprintf("encoding queueItem: %s", err),
+		}
+
+		replyErr := enc.Encode(reply)
+		if replyErr != nil {
+			return fmt.Errorf("encoding streamError reply: %w", replyErr)
+		}
+
+		return fmt.Errorf("encoding queueItem: %w", err)
+	}
+
+	err = enc.Encode(streamQueueReply{
+		Type: streamReplyDone,
+	})
+
+	if err != nil {
+		return fmt.Errorf("terminating stream: %w", err)
+	}
+
+	return nil
+}
+
+// debugIteratedOrdered will call f on each queueItem (sorted by indexing priority)
+// known to the queue, including queueItems that have been popped from the queue).
+//
+// Note: The mutex is held during all calls to f. Callers must not modify the queueItems.
+func (q *Queue) debugIteratedOrdered(f func(*queueItem)) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	queueItems := make([]*queueItem, len(q.items))
+
+	i := 0
+	for _, item := range q.items {
+		queueItems[i] = item
+		i++
+	}
+
+	sort.Slice(queueItems, func(i, j int) bool {
+		x, y := queueItems[i], queueItems[j]
+		return lessQueueItemPriority(x, y)
+	})
+
+	for _, item := range queueItems {
+		f(item)
+	}
+}
+
+type queueStreamReplyKind int
+
+const (
+	streamReplyQueueItem queueStreamReplyKind = iota
+	streamReplyDone
+	streamReplyError
+)
+
+type streamQueueReply struct {
+	Type        queueStreamReplyKind
+	Item        *queueItem
+	ErrorString string
+}
+
+func (q *Queue) handleDebugQueue(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		http.Error(w, "method must be GET", http.StatusBadRequest)
+		return
+	}
+
+	desiredContentType := r.Header.Get("Accept")
+	if desiredContentType != "application/x-gob-stream" {
```

Comment:
In the future we can `switch` off of this header to perhaps return JSON, HTML for a web version of this page, etc.